### ISSUE TITLE
Group release marklines in charts

### ISSUE THREAD

**Issue by gggritso** [getsentry/sentry#78041]
### Problem Statement

Many charts in Sentry have an option to show releases. Each release is marked with a blue vertical line. This is very handy, but for projects that cut a lot of releases (e.g., Sentry itself) this creates a _lot_ of blue lines. It's a bit distracting.

![Image](https://github.com/user-attachments/assets/3aabbeef-ea93-4749-b95e-c4883f6134af)

### Solution Brainstorm

Via @wmak, suggestion is to _group_ the lines, kind of like map pin clustering. Collapse lines into groups, where hovering over a group shows all releases in that group or bucket. Expand them upon zoom, as appropriate.

A naive approach is to partition the horizontal space into, say, 10 equal sections. For each section, if there is just one release in it, leave it alone. If there is more than one, plot a line at the average time coordinate of those releases. On hover, show a tooltip with every release in the group.

As uses zoom in, naturally more and more sections will have just a single release in them and grouping won't be needed. This works well but looks a little janky, with evenly spaced lines that don't represent the actual density of releases by time.

A less naive but harder approach is to dynamically group the releases by proximity to each other, with something like k-means. My guess is that even in one dimension k-means will be too slow because its time complexity is something obscene (at least exponential!) but it's pretty easy to implement and try. K-means also has the downside that the number of clusters is chosen upfront, but it's possible to change the implementation to add more groups if the spacing within each group is poor.

This gives a much better UI because the groups will naturally show up where there are a lot of releases.

I'm sure other approaches are possible!

**Comment by ryan953** [comment]
I think we should stay far away from any strategy that groups multiple lines into a single line. It would be very easy for a user to see a single line representing many releases and misinterpret the precision of that line within the graph. 

For example: look at this slice of the above graph, in the middle there are two lines roughly together (green arrow). They follow near a drop in the red metic:
![Image](https://github.com/user-attachments/assets/c0a33493-e68d-46ad-9a46-3a8aed1b01a2)

If we were to have grouped these then the position of the grouped line would change relative to the drop in the red metric. It'll either move to the left or right doesn't matter which..  just that the correlation becomes much weaker that the 1st release here might've had something to do with the metric drop.
Similarly, on the right side of this image there's a steep rise again correlating very strongly to the next release (pink arrow). 

It is reasonable for people to zoom in on the chart to see the precise release times overlaid on to the metric. This would work for line charts, as above. If this is the way users will use the system then it could be argued that we shouldn't show any release info across 90d (or 60d or 30d) timeframes because it's too noisy. I would buy that argument, but i think we can do better... 

I like the idea of grouping the release lines together, but i don't think we should use the same vertical-line treatment to show those groups of releases. Instead, whenever we have a chart I think we should be choosing between vertical lines or basically bar-graph style boxes for release based on:
| - | Metic uses line chart | Metric Uses bar Chart |
| --- | --- | --- |
| statsPeriod is small (<7?)  | releases use lines, no grouping  | release info available by hovering existing bars |
| statsPeriod is large (>7d?) | releases use bars, grouped | release info available by hovering bars |

Whenever we show releases grouped into a bar-chart style, the tooltips should be the way to access a summary of how many releases are in there, and deeplink to see that list.

❌  I think we shouldn't show precise release lines over bars like this:
![Image](https://github.com/user-attachments/assets/18f6c790-e2d8-4faf-aede-0aac3c8e0969)

✅  Instead we could show overlaid release data somehow, grouped to the same granularity as the chart (dora had a good mock of this, i'm doing a worse job here with the pink boxes):
![Image](https://github.com/user-attachments/assets/ae9a8a2f-0613-4bcf-aa74-babfd3874e31)

**Comment by vuluongj20** [comment]
I agree with Ryan here on the risk of misinterpretation. I haven't seen a design where grouped marks coexist with individual marks in an intuitive manner (but would love to see one if you have an example!)

I did some exploration into this problem a couple weeks ago. Like @ryan953 said, we should either show individual lines for each release or group all of them. When grouped, the binning interval should remain constant throughout the chart and match the interval of the events bars. The goal here is to make the grouping function more expectable.

![Image](https://github.com/user-attachments/assets/67f24b79-fd11-4c3e-99b0-838d8d91b5ed)
![Image](https://github.com/user-attachments/assets/1f5e1c9d-b2dc-4215-936a-0bc28b840d7d)

**Comment by ryan953** [comment]
Ah, it was you Vu who did the boxes underneath! i like those.

**Comment by gggritso** [comment]
🤔 👍🏻 

1. Sounds like mixing grouped and un-grouped releases can't work well, that's fair!
2. For bar charts, I agree that binning releases alongside the existing bars makes sense, and the mockup looks great
3. For line charts, I think overlaying grouped release bars over (or near) the line series might look strange, but that seems like a solveable UI problem
4. I don't think switching between individual lines and groups based on the statsPeriod duration will well. There could be very few releases in a long time frame, or the opposite. Decision to group or not group should be based on the count of releases vs. available space, I think
5. If we're definitely going to bin based on predictable intervals (rather than clustering) that makes the problem a lot easier 👍🏻

**Comment by billyvg** [comment]
As an update, the Replay team (in collab with Dashboards/Insights) will be tackling this issue in the upcoming quarter. This ticket is a bit too focused on the UI problem of having too many release lines on the chart. I'd like to re-frame the problem a bit to solve how users are actually using Releases in the context of charts. 

The most common use case is to correlate an anomaly on a chart with the Release (or feature flag) that caused it. The other use case I can think of (but have not confirmed the validity) is to check and verify the outcome of a release. e.g. I made a release and want to make sure nothing has regressed, or compare it vs previous release to see if there was an overall improvement.

We'll be brainstorming some more before settling on a solution.

**Comment by gggritso** [comment]
@billyvg 

> but have not confirmed the validity

It's valid, IMO. To me, an "anomaly" is anything that doesn't match expectations. A spike/drop where I expected steady state, or a steady state where I expected a spike/drop are both anomalies 👍🏻

**Comment by billyvg** [comment]
This is now completed! If you're using `TimeSeriesWidgetVisualization`, you can read [the storybook story](https://sentry.sentry.io/stories/?name=app%2Fviews%2Fdashboards%2Fwidgets%2FtimeSeriesWidget%2FtimeSeriesWidgetVisualization.stories.tsx&query=timew#releases) on how to use it.
